stratum/hal/lib/barefoot/bf_pipeline_utils.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright 2012-present Open Networking Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef STRATUM_HAL_LIB_BAREFOOT_BF_PIPELINE_UTIL_H_
+#define STRATUM_HAL_LIB_BAREFOOT_BF_PIPELINE_UTIL_H_
+
+#include "stratum/hal/lib/barefoot/bf_pipeline_utils.h"
+
+#include <string>
+#include <memory>
+#include <arpa/inet.h>
+#include <unistd.h>
+#include <vector>
+
+#include "absl/strings/strip.h"
+#include "libarchive/archive.h"
+#include "libarchive/archive_entry.h"
+#include "nlohmann/json.hpp"
+#include "p4/v1/p4runtime.pb.h"
+#include "stratum/glue/gtl/cleanup.h"
+#include "stratum/glue/status/status_macros.h"
+#include "stratum/lib/macros.h"
+#include "stratum/lib/utils.h"
+#include "stratum/hal/lib/barefoot/bf.pb.h"
+#include "stratum/public/lib/error.h"
+
+namespace stratum {
+namespace {
+// Helper function to extract the contents of first file named filename from
+// an in-memory archive.
+::util::StatusOr<std::string> ExtractFromArchive(const std::string& archive,
+                                                 const std::string& filename) {
+  struct archive* a = archive_read_new();
+  auto cleanup = gtl::MakeCleanup([&a]() { archive_read_free(a); });
+  archive_read_support_filter_bzip2(a);
+  archive_read_support_filter_xz(a);
+  archive_read_support_format_tar(a);
+  int r = archive_read_open_memory(a, archive.c_str(), archive.size());
+  CHECK_RETURN_IF_FALSE(r == ARCHIVE_OK) << "Failed to read archive";
+  struct archive_entry* entry;
+  while (archive_read_next_header(a, &entry) == ARCHIVE_OK) {
+    std::string path_name = archive_entry_pathname(entry);
+    if (absl::StripSuffix(path_name, filename) != path_name) {
+      VLOG(2) << "Found file: " << path_name;
+      std::string content;
+      content.resize(archive_entry_size(entry));
+      CHECK_RETURN_IF_FALSE(archive_read_data(a, &content[0], content.size()) ==
+                            content.size());
+      return content;
+    }
+  }
+  return MAKE_ERROR(ERR_ENTRY_NOT_FOUND) << "File not found: " << filename;
+}
+} // namespace
+
+namespace {
+// Helper function to convert a uint32 to a little-endian byte string
+std::string Uint32ToLeByteStream(uint32 val) {
+  uint32 tmp = (htonl(1) == 1) ? __builtin_bswap32(val) : val;
+  std::string bytes = "";
+  bytes.assign(reinterpret_cast<char*>(&tmp), sizeof(uint32));
+  return bytes;
+}
+} // namespace
+
+namespace hal {
+namespace barefoot {
+
+::util::Status ExtractBfDeviceConfig(
+    const ::p4::v1::ForwardingPipelineConfig& config,
+    BfPipelineConfig* bf_config) {
+
+  // FORMAT 1: p4_device_config is BfrtDeviceConfig instance
+  // Try a parse of BfrtDeviceConfig.
+  {
+    bf_config->Clear();
```

Comment:
Not needed: 

> The message is cleared first, even if the function fails 

https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.text_format#TextFormat.Parse.details